### ISSUE TITLE
Thread pool induced deadlocks and starvation

### ISSUE THREAD

**Issue by italovalcy** [kytos-ng/kytos#210]
Dear all,

I was running into issues when running Kytos in a medium-size environment [1] where the Kytos controller suddenly gets stuck and stops answering OpenFlow messages, switches get disconnected and never connect again, Napps stop handling events, etc. After adding a few extra log messages to the code [2], I could notice the following behavior which is very likely to be a deadlock/starvation induced by the thread pool way of managing the threads.

Quoting an [explanation from the web](http://asznajder.github.io/thread-pool-induced-deadlocks/):

> (...) Thread pool induced deadlocks may differ on the difficulty of finding their cause, but they most often follow the same schema. The simplest reason of deadlock is depending in one task (T1) on the results of another task (T2). If a thread pool is full and the second task (T2) gets queued, the first task (T1) might block and wait infinitely. Classic deadlock appears then: thread in a pool is occupied by the first task (T1) which waits for the second task (T2) that cannot be executed because all threads in a pool are busy. (...)

The first thing I noticed was the ThreadPoolExecutor queue size growing uncontrolled (output was summarized by minute to avoid very long output):
```
grep 'thread pool queue size' /var/log/kytos/kytos.log | uniq -w 16
2022-04-24 17:52:33,173 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 0
2022-04-24 17:53:00,001 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 0
2022-04-24 17:54:00,002 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 0
2022-04-24 17:55:00,023 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 0
2022-04-24 17:56:00,367 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 0
2022-04-24 17:57:00,456 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 0
2022-04-24 17:58:00,559 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 0
2022-04-24 17:59:00,003 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 0
2022-04-24 18:00:00,013 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 0
2022-04-24 18:01:00,849 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 419
2022-04-24 18:02:00,498 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 1118
2022-04-24 18:03:00,060 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 2196
2022-04-24 18:04:00,136 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 3276
2022-04-24 18:05:00,206 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 4356
2022-04-24 18:06:00,277 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 5427
2022-04-24 18:07:00,352 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 6507
2022-04-24 18:08:00,424 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 7585
2022-04-24 18:09:00,005 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 8644
2022-04-24 18:10:00,065 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 9722
2022-04-24 18:11:00,139 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 10802
2022-04-24 18:12:00,209 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 11882
2022-04-24 18:13:00,285 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 12962
...
2022-04-25 12:51:00,099 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 1116846
2022-04-25 12:52:00,172 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 1117926
2022-04-25 12:53:00,239 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 1119006
2022-04-25 12:54:00,312 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 1120086
2022-04-25 12:55:00,387 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 1121166
2022-04-25 12:56:00,459 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 1122246
```

Then, trying to understand what happened at the time thread pool queue size started growing, I've looked at events triggers right before that moment. After a few investigation, it come to my attention that most of them were related to Locks (threading Locks and Filelocks) and they looked just like a deadlock and starvation scenario:

One of the very previous messages logged before thread pool queue size start growing was this:

```
2022-04-24 19:34:14,553 - DEBUG [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_244) Lock trying to acquire lineno 265 uuid 05ecb018fed743a2aceba708826b129d
```

The log above was related to the Lock at Flow Manager `_publish_installed_flow()` `with self._storehouse_lock:` line. Then the question become: who was holding that Lock and what happened to that thread? The answer comes a few milliseconds before in the log:

```
2022-04-24 19:34:14,321 - DEBUG [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_147) Lock acquired lineno 267 uuid b24337dd58004cdfbcb99fe07a5612c8
```

Looking at the other part of the question (what happened to that thread?), this particular thread was also blocked into a Lock (but this time the `flow_manager.storehouse._lock` Lock) a few milliseconds after it acquires the `flow_manager._storehouse_lock` Lock:

```
2022-04-24 19:34:14,343 - DEBUG [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_147) save_flow (wait lock) namespace kytos.flow.persistence box_id c354ebb9bbea4638b720ce5906e66f56 uuid 19d9d648d5c8426ab1e50e5f896be003
```

The question then becomes: who was holding the `flow_manager.storehouse._lock ` Lock and what happened to that thread? The answer once again was a few milliseconds before at the log:

```
2022-04-24 19:34:14,225 - DEBUG [kytos.napps.kytos/flow_manager] (Thread-884) Lock <locked _thread.lock object at 0x7fe34e4e66c0> acquired.
```

Looking into the log messages, the Lock was never released:

```
root@d5d19a072126:/# awk '/^2022-04-24 19:34:14,225/,0' /var/log/kytos/kytos.log | grep '_thread.lock object at 0x7fe34e4e66c0'
2022-04-24 19:34:14,225 - DEBUG [kytos.napps.kytos/flow_manager] (Thread-884) Lock <locked _thread.lock object at 0x7fe34e4e66c0> acquired.
root@d5d19a072126:/#
```

What happened to this thread can be answered by looking at the sequence of log messages below (log entries below were manually organized to facilitate comprehension):

```
2022-04-24 19:34:14,225 - DEBUG [kytos.napps.kytos/flow_manager] (Thread-884) Lock <locked _thread.lock object at 0x7fe34e4e66c0> acquired.
2022-04-24 19:34:14,282 - DEBUG [kytos.napps.kytos/flow_manager] (Thread-884) storehouse.update namespace kytos.flow.persistence box_id c354ebb9bbea4638b720ce5906e66f56 uuid 3405366962674af3a7e7b87b52aa82ce
2022-04-24 19:34:14,289 - DEBUG [kytos.core.buffers] (Thread-884) [buffer: app_event] Added: kytos.storehouse.update
2022-04-24 19:34:14,380 - DEBUG [kytos.core.buffers] (MainThread) [buffer: app_event] Removed: kytos.storehouse.update
2022-04-24 19:34:14,385 - DEBUG [kytos.core.controller] (MainThread) looking for listeners for kytos.storehouse.update
```

The first line is the lock acquired from flow_manager.storehouse component. The second message is the debug message I've added just before calling `self.controller.buffers.app.put(event)`. The third is the Kytos core component informing that the message was added to KytosBuffer app_event queue. Then, the fourth and fifth messages are the Kytos core asyncio tasks that should create a thread and process the event in that thread. Indeed, I've added a log message right in the beginning of storehouse's handler for `kytos.storehouse.update `, however that log message was never seen after `looking for listeners for kytos.storehouse.update `:

```
root@d5d19a072126:/# awk '/^2022-04-24 19:34:14,385/,0' /var/log/kytos/kytos.log | grep storehouse.backend.update
root@d5d19a072126:/#
```

Why is that? I believe that is because the thread pool size (defaults to 256) is already full of requests in the waiting room and never gets the opportunity to create a thread to handle this request (which could unblock other threads that are actually blocking this one.. and there is where we have a deadlock/starvation). At that moment, all thread pool workers were allocated and the backlog/waiting-queue was already growing:

```
root@d5d19a072126:/# awk '/^2022-04-24 19:34:14,385/,0' /var/log/kytos/kytos.log | grep 'thread pool queue size'  | uniq -w 21 | head -n 20
2022-04-24 19:34:14,388 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 46
2022-04-24 19:34:14,402 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 47
2022-04-24 19:34:14,500 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 56
2022-04-24 19:34:15,385 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 56
2022-04-24 19:34:16,385 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 59
2022-04-24 19:34:17,388 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 65
2022-04-24 19:34:17,509 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 68
2022-04-24 19:34:17,602 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 76
2022-04-24 19:34:17,700 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 81
2022-04-24 19:34:18,389 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 84
2022-04-24 19:34:19,390 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 90
2022-04-24 19:34:20,391 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 93
2022-04-24 19:34:20,733 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 99
2022-04-24 19:34:20,803 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 106
2022-04-24 19:34:20,900 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 110
2022-04-24 19:34:21,393 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 115
2022-04-24 19:34:22,393 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 118
2022-04-24 19:34:23,394 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 124
2022-04-24 19:34:23,401 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 125
2022-04-24 19:34:23,961 - DEBUG [kytos.core.controller] (MainThread) thread pool queue size 127
```

[1] Description of the scenario

Kytos running into a docker container, using amlight/kytos:latest. Mininet topology was based on [AmlightTopo](https://github.com/amlight/kytos-end-to-end-tests/blob/master/tests/helpers.py#L10-L65) (`tmux new-sess -d -s mn mn --custom /e2e-tests/tests/helpers.py --topo amlight --controller=remote,ip=127.0.0.1`). There were 50 EVCs created between UNIs `00:00:00:00:00:00:00:16:55` and `00:00:00:00:00:00:00:21:60` (dynamic-path). Finally, to stress the system I've disabled one of the links in the path for the 50 EVCs (and mef_eline was running a modified version to process all EVC's link_down event in parallel).

The physical environment running the tests was a server with processor Intel(R) Xeon(R) CPU E5-2643 v3 @ 3.40GHz, 12 vcpus, 24GB of RAM memory.

[2] Extra log messages added to kytos (core and napps):

```
root@d5d19a072126:/src/kytos# git diff
diff --git a/kytos/core/controller.py b/kytos/core/controller.py
index c12a167..0e0da56 100644
--- a/kytos/core/controller.py
+++ b/kytos/core/controller.py
@@ -493,6 +493,8 @@ class Controller:
             event (~kytos.core.KytosEvent): An instance of a KytosEvent.
         """
         self.log.debug("looking for listeners for %s", event)
+        if executor_pool:
+            self.log.debug("thread pool queue size %d", executor_pool._work_queue.qsize())
         for event_regex, listeners in dict(self.events_listeners).items():
             # self.log.debug("listeners found for %s: %r => %s", event,
             #                event_regex, [l.__qualname__ for l in listeners])
root@d5d19a072126:/src/kytos-flow-manager# git diff
diff --git a/main.py b/main.py
index 911fbd1..d5fb498 100644
--- a/main.py
+++ b/main.py
@@ -7,6 +7,8 @@ from collections import OrderedDict, defaultdict
 from copy import deepcopy
 from enum import Enum
 from threading import Lock
+import inspect
+import uuid

 from flask import jsonify, request
 from napps.kytos.flow_manager.match import match_flow, match_strict_flow
@@ -154,8 +156,12 @@ class Main(KytosNApp):
         switch = event.content["source"].switch
         if not switch:
             return
+        my_uuid = uuid.uuid4().hex
+        log.debug("Lock trying to acquire lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
         with self._check_consistency_lock:
+            log.debug("Lock acquired lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
             self.reset_check_consistency_exec(switch.id)
+        log.debug("Lock done lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))

     def reset_check_consistency_exec(self, dpid):
         """Reset _check_consistency_exec_at for a dpid."""
@@ -214,20 +220,28 @@ class Main(KytosNApp):
         switch = event.source.switch
         message = event.message
         xid = int(message.header.xid)
+        my_uuid = uuid.uuid4().hex
+        log.debug("Lock trying to acquire lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
         with self._pending_barrier_lock:
+            log.debug("Lock acquired lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
             flow_xid = self._pending_barrier_reply[switch.id].pop(xid, None)
             if not flow_xid:
                 log.error(
                     f"Failed to pop barrier reply xid: {xid}, flow xid: {flow_xid}"
                 )
                 return
+        log.debug("Lock done lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
+        my_uuid = uuid.uuid4().hex
+        log.debug("Lock trying to acquire lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
         with self._flow_mods_sent_lock:
+            log.debug("Lock acquired lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
             if flow_xid not in self._flow_mods_sent:
                 log.error(
                     f"Flow xid: {flow_xid} not in flow mods sent. Inconsistent state"
                 )
                 return
             flow, _ = self._flow_mods_sent[flow_xid]
+        log.debug("Lock done lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))

         """
         It should only publish installed flow if it the original FlowMod xid hasn't
@@ -235,18 +249,26 @@ class Main(KytosNApp):
         hasn't been sent out or processed yet this can happen if the network latency
         is super low.
         """
+        my_uuid = uuid.uuid4().hex
+        log.debug("Lock trying to acquire lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
         with self._flow_mods_sent_error_lock:
+            log.debug("Lock acquired lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
             error_kwargs = self._flow_mods_sent_error.get(flow_xid)
+        log.debug("Lock done lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
         if not error_kwargs:
             self._publish_installed_flow(switch, flow)

     def _publish_installed_flow(self, switch, flow):
         """Publish installed flow when it's confirmed."""
         self._send_napp_event(switch, flow, "add")
+        my_uuid = uuid.uuid4().hex
+        log.debug("Lock trying to acquire lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
         with self._storehouse_lock:
+            log.debug("Lock acquired lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
             self._update_flow_state_store(
                 switch.id, set([flow.id]), FlowEntryState.INSTALLED.value
             )
+        log.debug("Lock done lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))

     @listen_to("kytos/of_core.flow_stats.received")
     def on_flow_stats_publish_installed_flows(self, event):
@@ -272,10 +294,14 @@ class Main(KytosNApp):
             flow_ids_to_update.add(_id)
             self._send_napp_event(switch, installed_flow, "add")

+        my_uuid = uuid.uuid4().hex
+        log.debug("Lock trying to acquire lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
         with self._storehouse_lock:
+            log.debug("Lock acquired lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
             self._update_flow_state_store(
                 switch.id, flow_ids_to_update, FlowEntryState.INSTALLED.value
             )
+        log.debug("Lock done lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))

     def _retry_on_openflow_connection_error(
         self,
@@ -309,7 +335,10 @@ class Main(KytosNApp):
             )
         switch = event.content["destination"].switch

+        my_uuid = uuid.uuid4().hex
+        log.debug("Lock trying to acquire lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
         with self._flow_mods_retry_count_lock:
+            log.debug("Lock acquired lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
             if xid not in self._flow_mods_retry_count:
                 self._flow_mods_retry_count[xid] = (0, now(), min_wait)
             (count, sent_at, wait_acc) = self._flow_mods_retry_count[xid]
@@ -327,6 +356,7 @@ class Main(KytosNApp):
                 datetime_t2,
                 wait_acc * multiplier,
             )
+        log.debug("Lock done lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
         try:
             wait_diff = get_min_wait_diff(datetime_t2, sent_at, wait_acc)
             if wait_diff:
@@ -350,7 +380,10 @@ class Main(KytosNApp):
         """Check consistency of stored and installed flows given a switch."""
         if not ENABLE_CONSISTENCY_CHECK or not switch.is_enabled():
             return
+        my_uuid = uuid.uuid4().hex
+        log.debug("Lock trying to acquire lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
         with self._check_consistency_lock:
+            log.debug("Lock acquired lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
             exec_at = self._check_consistency_exec_at.get(
                 switch.id, "0001-01-01T00:00:00"
             )
@@ -365,6 +398,7 @@ class Main(KytosNApp):
             self._check_consistency_exec_at[switch.id] = now().strftime(
                 "%Y-%m-%dT%H:%M:%S"
             )
+        log.debug("Lock done lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
         log.debug(f"check_consistency on switch {switch.id} has started")
         self.check_storehouse_consistency(switch)
         if switch.dpid in self.stored_flows:
@@ -615,7 +649,10 @@ class Main(KytosNApp):
         """Add archived flows in memory (will be stored in the DB in the future)."""
         if not archived_flows:
             return
+        my_uuid = uuid.uuid4().hex
+        log.debug("Lock trying to acquire lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
         with self._archived_flows_lock:
+            log.debug("Lock acquired lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
             if dpid not in self.archived_flows:
                 self.archived_flows[dpid] = OrderedDict()
             for archived_flow in archived_flows:
@@ -623,6 +660,7 @@ class Main(KytosNApp):

             while len(self.archived_flows[dpid]) > max_len:
                 self.archived_flows[dpid].popitem(last=False)
+        log.debug("Lock done lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))

     def _update_flow_state_store(self, dpid, flow_ids, state):
         """Try to bulk update the state of some flow ids given a dpid."""
@@ -866,14 +904,22 @@ class Main(KytosNApp):
                 except SwitchNotConnectedError:
                     if reraise_conn:
                         raise
+                my_uuid = uuid.uuid4().hex
+                log.debug("Lock trying to acquire lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
                 with self._flow_mods_sent_lock:
+                    log.debug("Lock acquired lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
                     self._add_flow_mod_sent(flow_mod.header.xid, flow, command)
+                log.debug("Lock done lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
                 self._send_napp_event(switch, flow, "pending")

                 if not save:
                     continue
+                my_uuid = uuid.uuid4().hex
+                log.debug("Lock trying to acquire lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
                 with self._storehouse_lock:
+                    log.debug("Lock acquired lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
                     self._store_changed_flows(command, flow_dict, flow.id, switch)
+                log.debug("Lock done lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))

     def _add_flow_mod_sent(self, xid, flow, command):
         """Add the flow mod to the list of flow mods sent."""
@@ -896,8 +942,12 @@ class Main(KytosNApp):

         barrier_request = new_barrier_request(switch.connection.protocol.version)
         barrier_xid = barrier_request.header.xid
+        my_uuid = uuid.uuid4().hex
+        log.debug("Lock trying to acquire lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
         with self._pending_barrier_lock:
+            log.debug("Lock acquired lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
             self._add_barrier_request(switch.id, barrier_xid, flow_mod.header.xid)
+        log.debug("Lock done lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))

         content = {"destination": switch.connection, "message": barrier_request}
         event = KytosEvent(name=event_name, content=content)
@@ -953,8 +1003,12 @@ class Main(KytosNApp):
             "error_command": error_command,
             "error_exception": event.content.get("exception"),
         }
+        my_uuid = uuid.uuid4().hex
+        log.debug("Lock trying to acquire lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
         with self._flow_mods_sent_error_lock:
+            log.debug("Lock acquired lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
             self._flow_mods_sent_error[int(event.message.header.xid)] = error_kwargs
+        log.debug("Lock done lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
         self._send_napp_event(
             switch,
             flow,
@@ -1013,12 +1067,20 @@ class Main(KytosNApp):
                 "error_type": error_type,
                 "error_code": error_code,
             }
+            my_uuid = uuid.uuid4().hex
+            log.debug("Lock trying to acquire lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
             with self._flow_mods_sent_error_lock:
+                log.debug("Lock acquired lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
                 self._flow_mods_sent_error[int(event.message.header.xid)] = error_kwargs
+            log.debug("Lock done lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
             log.warning(
                 f"Deleting flow: {flow.as_dict()}, xid: {xid}, cookie: {flow.cookie}, "
                 f"error: {error_kwargs}"
             )
+            my_uuid = uuid.uuid4().hex
+            log.debug("Lock trying to acquire lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
             with self._storehouse_lock:
+                log.debug("Lock acquired lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
                 self._del_stored_flow_by_id(switch.id, flow.cookie, flow.id)
+            log.debug("Lock done lineno %d uuid %s" % (inspect.stack()[0].lineno, my_uuid))
             self._send_napp_event(flow.switch, flow, "error", **error_kwargs)
diff --git a/storehouse.py b/storehouse.py
index 81d32c7..8e33c0f 100644
--- a/storehouse.py
+++ b/storehouse.py
@@ -1,6 +1,7 @@
 """Module to handle the storehouse."""
 import threading
 import time
+import uuid

 from napps.kytos.flow_manager import settings

@@ -91,16 +92,20 @@ class StoreHouse:
     # pylint: disable=consider-using-with
     def get_stored_box(self, box_id):
         """Get persistence box from storehouse."""
+        my_uuid = uuid.uuid4().hex
         content = {
             "namespace": self.namespace,
             "callback": self._get_box_callback,
             "box_id": box_id,
+            'my_uuid': my_uuid,
             "data": {},
         }
         name = "kytos.storehouse.retrieve"
         event = KytosEvent(name=name, content=content)
+        log.debug("get_stored_box (wait lock) namespace %s box_id %s uuid %s" % (self.namespace, box_id, my_uuid))
         self._lock.acquire()  # Lock to avoid race condition
         log.debug(f"Lock {self._lock} acquired.")
+        log.debug("storehouse.retrieve namespace %s box_id %s uuid %s" % (self.namespace, box_id, my_uuid))
         self.controller.buffers.app.put(event)

     def _get_box_callback(self, _event, data, error):
@@ -115,6 +120,8 @@ class StoreHouse:
     # pylint: disable=consider-using-with
     def save_flow(self, flows):
         """Save flows in storehouse."""
+        my_uuid = uuid.uuid4().hex
+        log.debug("save_flow (wait lock) namespace %s box_id %s uuid %s" % (self.namespace, self.box.box_id, my_uuid))
         self._lock.acquire()  # Lock to avoid race condition
         log.debug(f"Lock {self._lock} acquired.")
         self.box.data[flows["id"]] = flows
@@ -122,10 +129,12 @@ class StoreHouse:
             "namespace": self.namespace,
             "box_id": self.box.box_id,
             "data": self.box.data,
+            'my_uuid': my_uuid,
             "callback": self._save_flow_callback,
         }

         event = KytosEvent(name="kytos.storehouse.update", content=content)
+        log.debug("storehouse.update namespace %s box_id %s uuid %s" % (self.namespace, self.box.box_id, my_uuid))
         self.controller.buffers.app.put(event)

     def _save_flow_callback(self, _event, data, error):
root@d5d19a072126:/src/kytos-storehouse# git diff
diff --git a/main.py b/main.py
index 42cbd87..cba06b7 100644
--- a/main.py
+++ b/main.py
@@ -291,7 +291,9 @@ class Main(KytosNApp):
     def event_retrieve(self, event):
         """Retrieve a box from a namespace based on an event."""
         error = None
+        my_uuid = event.content.get('my_uuid', uuid4().hex)

+        log.debug("storehouse.backend.retrieve namespace %s box_id %s uuid %s" % (event.content['namespace'], event.content['box_id'], my_uuid))
         try:
             box = self.backend.retrieve(event.content['namespace'],
                                         event.content['box_id'])
@@ -299,6 +301,7 @@ class Main(KytosNApp):
             box = None
             error = exc

+        log.debug("storehouse.backend.retrieve done (exec callback) namespace %s box_id %s uuid %s" % (event.content['namespace'], event.content['box_id'], my_uuid))
         self._execute_callback(event, box, error)

     @listen_to('kytos.storehouse.update')
@@ -313,7 +316,9 @@ class Main(KytosNApp):
         data: a python dict with the data
         """
         error = False
+        my_uuid = event.content.get('my_uuid', uuid4().hex)

+        log.debug("storehouse.backend.update namespace %s box_id %s uuid %s" % (event.content['namespace'], event.content['box_id'], my_uuid))
         try:
             namespace = event.content['namespace']
             box_id = event.content['box_id']
@@ -335,6 +340,7 @@ class Main(KytosNApp):

             self.backend.update(namespace, box)

+        log.debug("storehouse.backend.update done (exec callback) namespace %s box_id %s uuid %s" % (event.content['namespace'], event.content['box_id'], my_uuid))
         self._execute_callback(event, box, error)

     @listen_to('kytos.storehouse.delete')
```

**Comment by italovalcy** [comment]
Running Kytos in the standard Thread (non thread pool) strategy seems to workaround for this issue:

```
tmux new-session -d -s kytosserver "kytosd -E -f -t 0"
```

**Comment by viniarck** [comment]
@italovalcy thanks for reporting this and this helpful debug analysis, very interesting and I'm glad this was caught at this stage. 

I believe storehouse async lock acquire/release on different thread is the only situation that is vulnerable to reach to this major starvation issue since on other `listen_to` decorators the lock is held with a context manager. I think the solution will be multifold and it might have some overlap with priority queues (I mean we'll probably want to have a solution aligned to that too). 

1 - We could have multiple ThreadPoolExecutors where each should be used accordingly depending on the priority or how critical the category of the event is, for instance we could seggregate into:

- `control`: for control plane related stuff, but aren't primarily IO-bound
- `io`: for handlers that are IO-bound
- `napps`: for anything else (default)

That way we could parametrize accordingly in the decorator the category of that handler with a `pool` parameter or something similar, for instance: 

`@listen_to("kytos/of_core.v0x0[14].messages.in.ofpt_barrier_reply", pool="io")`

It's probably worth mentioning that as we migrate to Mongo, the number of locks will also tend to reduce (including that async aquire/release), however, we probably want to make sure that IO/control bound handlers still have higher priority.

2 - We might also need to reassess the default number of workers in the default thread pool depending if we have additional ones. 


Out of curiosity, is it reproducing deterministically with this topology? Let me know also what you think about these initial ideas.

**Comment by italovalcy** [comment]
@viniarck I've tried to modify flow_manager's storehouse to use context based locks but so far no good outcomes. Here are more details:

First of all, below is the modification I made in mef_eline to process multiple EVCs in parallel (as events) upon link down message:

```
root@7434f28148ff:/src/kytos-mef-eline# git diff
diff --git a/main.py b/main.py
index 3a59a6b..4569a58 100644
--- a/main.py
+++ b/main.py
@@ -639,23 +639,45 @@ class Main(KytosNApp):

     def handle_link_down(self, event):
         """Change circuit when link is down or under_mantenance."""
-        log.debug("Event handle_link_down %s", event)
+        log.info("Event handle_link_down %s", event)
         for evc in self.circuits.values():
-            with evc.lock:
-                if evc.is_affected_by_link(event.content["link"]):
-                    log.debug(f"Handling evc {evc.id} on link down")
-                    if evc.handle_link_down():
-                        emit_event(
-                            self.controller,
-                            "redeployed_link_down",
-                            evc_id=evc.id,
-                        )
-                    else:
-                        emit_event(
-                            self.controller,
-                            "error_redeploy_link_down",
-                            evc_id=evc.id,
-                        )
+            if evc.is_affected_by_link(event.content["link"]):
+                emit_event(
+                    self.controller,
+                    "evc_affected_by_link_down",
+                    evc_id=evc.id,
+                    link=event.content["link"],
+                )
+
+    @listen_to("kytos/mef_eline.evc_affected_by_link_down")
+    def on_evc_affected_by_link_down(self, event):
+        """Change circuit when link is down or under_mantenance."""
+        self.handle_evc_affected_by_link_down(event)
+
+    def handle_evc_affected_by_link_down(self, event):
+        try:
+            evc = self.circuits[event.content["evc_id"]]
+        except KeyError:
+            result = f"circuit_id {circuit_id} not found"
+            log.error("handle evc affected by link down - invalid circuit_id"
+                      f" {circuit_id}")
+
+        log.info(f"Handling evc {evc.id} on link down")
+        # TODO: if link is already up here, can we just ignore this event?
+        with evc.lock:
+            if evc.handle_link_down():
+                emit_event(
+                    self.controller,
+                    "redeployed_link_down",
+                    evc_id=evc.id,
+                )
+                log.info(f"EVC {evc.id} redeployed due to link down")
+            else:
+                emit_event(
+                    self.controller,
+                    "error_redeploy_link_down",
+                    evc_id=evc.id,
+                )

     def load_all_evcs(self):
         """Try to load all EVCs on startup."""
```

Change I've made to flow_manager's storehouse to use context based locks:

```
diff --git a/settings.py b/settings.py
index 6b904c3..b48c1ad 100644
--- a/settings.py
+++ b/settings.py
@@ -20,3 +20,10 @@ ARCHIVED_FLOWS_MAX_SIZE = 5000
 CONN_ERR_MAX_RETRIES = 3
 CONN_ERR_MIN_WAIT = 1  # minimum wait between iterations in seconds
 CONN_ERR_MULTIPLIER = 2  # multiplier for the accumulated wait on each iteration
+
+# Time (in seconds) to wait for a confirmation from storehouse
+# when retrieving or updating a box
+STOREHOUSE_TIMEOUT = 5.0
+
+# Time (in seconds) to sleep while waiting from storehouse (busy wait)
+STOREHOUSE_WAIT_INTERVAL = 0.05
diff --git a/storehouse.py b/storehouse.py
index 81d32c7..d248385 100644
--- a/storehouse.py
+++ b/storehouse.py
@@ -36,15 +36,17 @@ class StoreHouse:
             self.box = None
         if "_lock" not in self.__dict__:
             self._lock = threading.Lock()
+        self.storehouse_done = False
+        self.storehouse_error = None
         self.list_stored_boxes()

     def get_data(self):
         """Return the persistence box data."""
         # Wait retrieve or create box in storehouse
         i = 0
-        while not self.box and i < BOX_RESTORE_ATTEMPTS:
-            time.sleep(self.box_restore_timer)
-            i += 1
+        while not self.box and i < settings.STOREHOUSE_TIMEOUT:
+            time.sleep(settings.STOREHOUSE_WAIT_INTERVAL)
+            i += settings.STOREHOUSE_WAIT_INTERVAL
         if not self.box:
             error = "Error retrieving persistence box from storehouse."
             log.error(error)
@@ -91,48 +93,69 @@ class StoreHouse:
     # pylint: disable=consider-using-with
     def get_stored_box(self, box_id):
         """Get persistence box from storehouse."""
-        content = {
-            "namespace": self.namespace,
-            "callback": self._get_box_callback,
-            "box_id": box_id,
-            "data": {},
-        }
-        name = "kytos.storehouse.retrieve"
-        event = KytosEvent(name=name, content=content)
-        self._lock.acquire()  # Lock to avoid race condition
-        log.debug(f"Lock {self._lock} acquired.")
-        self.controller.buffers.app.put(event)
-
-    def _get_box_callback(self, _event, data, error):
+        with self._lock:
+            self.storehouse_done = False
+            self.storehouse_error = None
+            content = {
+                "namespace": self.namespace,
+                "callback": self._storehouse_callback,
+                "box_id": box_id,
+                "data": {},
+            }
+            name = "kytos.storehouse.retrieve"
+            event = KytosEvent(name=name, content=content)
+            self.controller.buffers.app.put(event)
+
+            i = 0
+            while not self.storehouse_done and i < settings.STOREHOUSE_TIMEOUT:
+                time.sleep(settings.STOREHOUSE_WAIT_INTERVAL)
+                i += settings.STOREHOUSE_WAIT_INTERVAL
+
+            if not self.storehouse_done:
+                self.storehouse_error = 'Timeout while waiting for storehouse'
+
+            if self.storehouse_error:
+                log.error('Fail to update persistence box in '
+                          f'{self.namespace}.{self.box.box_id}. '
+                          f'Error: {self.storehouse_error}')
+                return
+
+        log.debug(f"get_stored_box done {self.namespace}.{self.box.box_id}")
+
+    def _storehouse_callback(self, _event, data, error):
         """Handle get_box method saving the box or logging with the error."""
-        if error:
-            log.error("Persistence box not found.")
-
-        self.box = data
-        self._lock.release()
-        log.debug(f"Lock {self._lock} released.")
+        self.storehouse_done = True
+        self.storehouse_error = error

     # pylint: disable=consider-using-with
     def save_flow(self, flows):
         """Save flows in storehouse."""
-        self._lock.acquire()  # Lock to avoid race condition
-        log.debug(f"Lock {self._lock} acquired.")
-        self.box.data[flows["id"]] = flows
-        content = {
-            "namespace": self.namespace,
-            "box_id": self.box.box_id,
-            "data": self.box.data,
-            "callback": self._save_flow_callback,
-        }
-
-        event = KytosEvent(name="kytos.storehouse.update", content=content)
-        self.controller.buffers.app.put(event)
-
-    def _save_flow_callback(self, _event, data, error):
-        """Display stored flow."""
-        self._lock.release()
-        log.debug(f"Lock {self._lock} released.")
-        if error:
-            log.error(f"Can't update persistence box {data.box_id}.")
-
-        log.info(f"Flow saved in {self.namespace}.{data.box_id}")
+        with self._lock:
+            self.storehouse_done = False
+            self.storehouse_error = None
+            self.box.data[flows["id"]] = flows
+            content = {
+                "namespace": self.namespace,
+                "box_id": self.box.box_id,
+                "data": self.box.data,
+                "callback": self._storehouse_callback,
+            }
+
+            event = KytosEvent(name="kytos.storehouse.update", content=content)
+            self.controller.buffers.app.put(event)
+
+            i = 0
+            while not self.storehouse_done and i < settings.STOREHOUSE_TIMEOUT:
+                time.sleep(settings.STOREHOUSE_WAIT_INTERVAL)
+                i += settings.STOREHOUSE_WAIT_INTERVAL
+
+            if not self.storehouse_done:
+                self.storehouse_error = 'Timeout while waiting for storehouse'
+
+            if self.storehouse_error:
+                log.error('Fail to update persistence box in '
+                          f'{self.namespace}.{self.box.box_id}. '
+                          f'Error: {self.storehouse_error}')
+                return
+
+        log.info(f"Flow saved in {self.namespace}.{self.box.box_id}")
```

Steps to reproduce are basically:

1. Run Kytos docker imagem from docker hub: `docker run -d --name k2 --privileged -v $(pwd)/kytos-end-to-end-tests:/e2e-tests -w /e2e-tests amlight/kytos`
2. Run mininet with Amlight's topology: `tmux new-sess -d -s mn mn --custom /e2e-tests/tests/helpers.py --topo amlight --controller=remote,ip=127.0.0.1`
3. Create a few EVCs with dynamic path from certain UNIs:
```
root@7434f28148ff:/e2e-tests# cat create-evc-dyn.sh
#!/bin/bash

IFACEID_A=00:00:00:00:00:00:00:16:55
IFACEID_Z=00:00:00:00:00:00:00:21:60

VLAN_A=$1
VLAN_Z=$2
test -z "$VLAN_Z" && VLAN_Z=$VLAN_A

for VLAN in $(seq $VLAN_A $VLAN_Z); do
	curl --max-time 120 -H 'Content-type: application/json' -X POST http://127.0.0.1:8181/api/kytos/mef_eline/v2/evc -d '{"name": "my evc '$VLAN'", "dynamic_backup_path": true, "enabled": true, "uni_a": {"tag": {"value": '$VLAN', "tag_type": 1}, "interface_id": "'$IFACEID_A'"}, "uni_z": {"tag": {"value": '$VLAN', "tag_type": 1}, "interface_id": "'$IFACEID_Z'"}}'
done
root@7434f28148ff:/e2e-tests# ./create-evc-dyn.sh 400 449
...
```
4. Access Mininet and simulate a link down event: `tmux send-keys -t mn "sh ip link set up AL2-eth6" ENTER`

**Comment by viniarck** [comment]
@italovalcy appreciated your help and investigation, thanks also for the topology you shared, it facilitated to reproduce the issue, the fix is under review on PR #213.
